Fix symlink resolution

ajeddeloh opened the pull request:
Fixes https://github.com/coreos/ignition/issues/640

arithx commented on tests/positive/files/link.go:
```diff
@@ -122,6 +124,105 @@ func CreateSymlinkOnRoot() types.Test {
+	name := "Write Through Relative Symlink on the Root Filesystem"
+	in := types.GetBaseDisk()
+	out := types.GetBaseDisk()
+	// note this abuses the order in which ignition writes links and will break with 3.0.0
```
Might be worthwhile to add a `TODO:` for easy grepping when branching for 3.0.0

ajeddeloh replied:
I could go either way. We'll need to audit the tests for 3.0.0 anyway and if it breaks we'll see this note.

cgwalters commented:
One thing we could do at least in rpm-ostree is force all symlinks to be relative.

cgwalters commented:
I'm guessing it'd be a whole lot of work, but having Ignition process filesystem changes inside a container would fix this too.  Or, maybe simpler, a bit like what jlebon did in rpm-ostree, spawn a helper binary in the target root to resolve the path?

There's a whole lot potentially involved in pathname resolution; you've got a simple implementation which I think is probably OK for the simple cases.  It will happily for example walk off the root if there are too many `../` though for example, but those symlinks will work fine since `../` in `/` is still `/`.

ajeddeloh commented:
>  It will happily for example walk off the root if there are too many ../ 

Damn, yeah I need to fix that.

> I'm guessing it'd be a lot of work, but having Ignition process filesystem changes inside a container would fix this too. Or, maybe simpler, a bit like what jlebon did in rpm-ostree, spawn a helper binary in the target root to resolve the path?

I was originally going to use `realpath` in the root but that hits issues where not every root will be a real rootfs, some will be empty. In that case we'd need to copy a static binary in, run it, then delete it. Unless you have some container magic that can handle that better.

bgilbert commented on internal/exec/stages/files/filesystemEntries.go:
```diff
@@ -101,12 +101,15 @@ func (tmp dirEntry) create(l *log.Logger, u util.Util) error {
 		if err != nil {
 			return err
 		}
+		path = filepath.Clean(path)
```
Should be the responsibility of `util.JoinPath`.

ajeddeloh commented:
Pushed an update. Think I got it right this time.

ajeddeloh commented:
It shouldn't be able to escape, but I can add the check back in.

ajeddeloh commented:
Ok, pushed hopefully the last update. It now follows symlinks as if rooted in the root filesystem for _only_ the rootfs and errors out on other ones. Added tests for both cases and added a section in the operator notes.

bgilbert commented on internal/exec/util/util.go:
```diff
@@ -15,20 +15,81 @@
+				return "", errEscapedMountpoint
+			}
+		}
+		if !u.IsRoot && wantsToEscape(symlinkPath) {
```
Putting this in an `else` or `else if` of the `IsAbs` block might be clearer.

Also, it might be good to add a comment that in the `u.IsRoot` case, the link will still not be able to escape `u.DestDir`.

bgilbert commented on tests/validator.go:
```diff
@@ -289,7 +289,7 @@ func validateLink(t *testing.T, partition *types.Partition, link types.Link) {
 		}
 		if targetPath != link.Target {
-			t.Error("Actual and expected symbolic link targets don't match")
+			t.Errorf("Actual and expected symbolic link targets don't match. Exepcted %q, got %q", link.Target, targetPath)
```
`expected`

bgilbert commented on doc/operator-notes.md:
```diff
@@ -31,6 +31,10 @@ If `wipeFilesystem` is set to true, Ignition will always wipe any preexisting fi
 
 If `wipeFilesystem` is set to false, Ignition will then attempt to reuse the existing filesystem. If the filesystem is of the correct type, has a matching label, and has a matching UUID, then Ignition will reuse the filesystem. If the label or UUID is not set in the Ignition config, they don't need to match for Ignition to reuse the filesystem. Any preexisting data will be left on the device and  ...
 
+## Path Traversal and Following symlnks
```
`Symlinks`

bgilbert commented on doc/operator-notes.md:
```diff
@@ -31,6 +31,10 @@ If `wipeFilesystem` is set to true, Ignition will always wipe any preexisting fi
 
+## Path Traversal and Following symlnks
+
+When resolving paths, Ignition follows symlinks on all but the last element of a path. This ensures symlinks already existing on a filesystem can be overwritten while still following symlinks as expected. When writing files, links, or directories, Ignition does not allow following symlinks outside the specified filesystem. When writing a file, link, or directory, on the `root` filesystem Ignition ...
```
commas: `When writing a file, link, or directory on the `root` filesystem,`

Mix of present and future tense.

ajeddeloh replied:
I think the tense mixing makes sense. It's both describing what Ignition will do and the rules which Ignition follows. I think it would definitely be awkward to do it all in the future tense (`Ignition will disallow` sounds weird) and it's easier to understand the cause/effect if the future tense is used for other sections.
